File: admin-sdk/v1.5.0-dev.34/javascript/webhooks/src/signing.js
'use strict';

// HMAC scheme shared with the MediaViz producer: X-Signature = 'sha256=' +
// hex(HMAC-SHA256(secret, `${X-Timestamp}.` + rawBody)). The signing string is
// the timestamp joined to the *raw request body bytes* — never re-serialized
// JSON. Verification accepts the current and (during rotation) previous
// secret, rejects timestamps outside the skew window, and compares digests in
// constant time. Web Crypto only, so the same code runs in Node >= 18 and
// browsers.

const _encoder = new TextEncoder();

async function signWebhookPayload(secret, timestamp, rawBody) {
  const key = await globalThis.crypto.subtle.importKey(
    'raw', _encoder.encode(secret), { name: 'HMAC', hash: 'SHA-256' }, false, ['sign']);
  const payload = _concat(_encoder.encode(`${timestamp}.`), toBytes(rawBody));
  const digest = await globalThis.crypto.subtle.sign('HMAC', key, payload);
  return 'sha256=' + _hex(new Uint8Array(digest));
}

async function verifyWebhookSignature(secretCurrent, secretPrevious, headers, rawBody, options = {}) {
  const h = lowerHeaders(headers);
  const timestamp = h['x-timestamp'];
  const signature = h['x-signature'];
  if (!timestamp || !signature) return false;
  if (!/^-?\d+$/.test(String(timestamp))) return false;
  const skewToleranceS = options.skewToleranceS ?? 300;
  const now = options.now ?? Math.floor(Date.now() / 1000);
  if (Math.abs(now - Number(timestamp)) > skewToleranceS) return false;
  for (const secret of [secretCurrent, secretPrevious]) {
    if (!secret) continue;
    const expected = await signWebhookPayload(secret, String(timestamp), rawBody);
    if (_constantTimeEqual(expected, String(signature))) return true;
  }
  return false;
}

// helpers
function toBytes(rawBody) {
  if (typeof rawBody === 'string') return _encoder.encode(rawBody);
  if (rawBody instanceof Uint8Array) return rawBody;
  if (rawBody instanceof ArrayBuffer) return new Uint8Array(rawBody);
  throw new TypeError('rawBody must be a string, Uint8Array, or ArrayBuffer');
}

function lowerHeaders(headers) {
  const out = {};
  const entries = typeof headers?.entries === 'function' ? headers.entries() : Object.entries(headers ?? {});
  for (const [k, v] of entries) out[String(k).toLowerCase()] = v;
  return out;
}

function _concat(a, b) {
  const out = new Uint8Array(a.length + b.length);
  out.set(a, 0);
  out.set(b, a.length);
  return out;
}

function _hex(bytes) {
  return Array.from(bytes, (b) => b.toString(16).padStart(2, '0')).join('');
}

function _constantTimeEqual(a, b) {
  if (a.length !== b.length) return false;
  let diff = 0;
  for (let i = 0; i < a.length; i++) diff |= a.charCodeAt(i) ^ b.charCodeAt(i);
  return diff === 0;
}

module.exports = { signWebhookPayload, verifyWebhookSignature, toBytes, lowerHeaders };

File: webhook_module/python/mediaviz_webhooks/_signing.py
"""HMAC scheme shared with the MediaViz producer.

X-Signature = 'sha256=' + hex(HMAC-SHA256(secret, f"{X-Timestamp}." + raw_body)).
The signing string is the timestamp joined to the *raw request body bytes* —
never re-serialized JSON. Verification accepts the current and (during
rotation) previous secret, rejects timestamps outside the skew window, and
compares digests in constant time.
"""
from __future__ import annotations

import hashlib
import hmac
import time


def sign_webhook_payload(secret: str, timestamp: str, raw_body: bytes | str) -> str:
    digest = hmac.new(secret.encode(), f"{timestamp}.".encode() + _to_bytes(raw_body), hashlib.sha256)
    return "sha256=" + digest.hexdigest()


def verify_webhook_signature(
    secret_current: str | None,
    secret_previous: str | None,
    headers: dict,
    raw_body: bytes | str,
    *,
    skew_tolerance_s: int = 300,
    now: float | None = None,
) -> bool:
    h = {str(k).lower(): v for k, v in dict(headers).items()}
    timestamp, signature = h.get("x-timestamp"), h.get("x-signature")
    if not timestamp or not signature:
        return False
    try:
        ts = int(timestamp)
    except (TypeError, ValueError):
        return False
    if abs((time.time() if now is None else now) - ts) > skew_tolerance_s:
        return False
    for secret in (secret_current, secret_previous):
        if secret and hmac.compare_digest(sign_webhook_payload(secret, str(timestamp), raw_body), str(signature)):
            return True
    return False


def _to_bytes(raw_body: bytes | str) -> bytes:
    if isinstance(raw_body, bytes):
        return raw_body
    if isinstance(raw_body, str):
        return raw_body.encode()
    raise TypeError("raw_body must be bytes or str")
